### PR TITLE
[model weights] zero_to_fp32 multiple improvements

### PR DESCRIPTION
**Pull request by stas00**
As requested at https://github.com/microsoft/DeepSpeed/issues/1062#issuecomment-865994154 for small models it's too awkward to go hunt for the checkpoint folder, run commands manually and go back to the software to use it.

So this PR:

- [x] provides a similar API to do it on the fly:
```
from deepspeed.utils.zero_to_fp32 import get_fp32_state_dict_from_zero_checkpoint
state_dict = get_fp32_state_dict_from_zero_checkpoint(checkpoint_dir) # already on cpu
model = model.cpu()
model.load_state_dict(state_dict)
#submit to model hub or save the model to share with others
```
so all you need here is the path to the pytorch checkpoint dir.

The only possible culprit here is that the model will no longer be useable in the deepspeed context of the same application. i.e. you will need to re-init the deepspeed engine, since `model.load_state_dict(state_dict)` will remove all the deepspeed magic from it.

And of course the main concern here is that the application will not have enough memory to accommodate this process.

- [x] provides an even simpler API to do it on the fly:

```
from deepspeed.utils.zero_to_fp32 import load_state_dict_from_zero_checkpoint
model = load_state_dict_from_zero_checkpoint(trainer.model, checkpoint_dir)
# now save it and push to hub
# note the model is no longer deepspeed-usable and requires a new deepspeed initialize if you want to continue using it right away
```

- [x] Also note, I changed the API to find the deepspeed sub-dir automatically so all you need is the path to the best or last normal checkpoint. i.e. you don't need the `global_step2` in the path.

Here are a few additional unrelated fixes:
- [x] in stage3 added a detector for params that didn't get partitioned
- [x] rewrote `_get_zero_param_shapes` to rely on `optimizer.fp16_groups` as `module.named_parameters` was completely unreliable
- [x] added a fix to copy `zero_to_fp32.py` only once
- [x] added a check that `len(optim_files)==world_size` - had a user who was missing one optim file.
- [x] multiple other sanity checks and validations and easier debug 

@piegu, please give it a try and let me know if it makes things easy for you.

I'm totally open to any suggestions to make it better.

@samyam, @jeffra, @tjruwase

### CODE REVIEW

**Comment by piegu** [comment]
Hello @stas00.

Thanks for your proposal. I updated my notebook with it but it seems there is a path problem to a file.
Link to notebook: https://colab.research.google.com/drive/17vN1w0HxRMopzps8WrlRwMmHhZppVQli?usp=sharing`

Code
````
from deepspeed.utils.zero_to_fp32 import get_fp32_state_dict_from_zero_chkpt
state_dict = get_fp32_state_dict_from_zero_chkpt(checkpoint_dir) # already on cpu
model = model.cpu()
model.load_state_dict(state_dict)
````

Result

```
Processing zero checkpoint '/content/test-mlm/checkpoint-162/global_step162'

---------------------------------------------------------------------------

FileNotFoundError                         Traceback (most recent call last)

<ipython-input-53-b8e677ab1e2f> in <module>()
      1 from deepspeed.utils.zero_to_fp32 import get_fp32_state_dict_from_zero_chkpt
----> 2 state_dict = get_fp32_state_dict_from_zero_chkpt(checkpoint_dir) # already on cpu
      3 model = model.cpu()
      4 model.load_state_dict(state_dict)
      5 #submit to model hub or save the model to share with others

2 frames

/usr/local/lib/python3.7/dist-packages/deepspeed/utils/zero_to_fp32.py in get_model_state_file(checkpoint_dir)
     29 
     30     if not os.path.exists(file):
---> 31         raise FileNotFoundError(f"can't find '{file}' in directory '{checkpoint_dir}'")
     32 
     33     return file

FileNotFoundError: can't find '/content/test-mlm/checkpoint-162/global_step162/zero_pp_rank_0_mp_rank_00_model_states.pt' in directory '/content/test-mlm/checkpoint-162/global_step162'
```

**Comment by stas00** [comment]
colab is extremely slow, got it running but meanwhile you could also check the contents of `/content/test-mlm/checkpoint-162/global_step162/` if you still have it open

**Comment by piegu** [comment]
> but meanwhile you could also check the contents of `/content/test-mlm/checkpoint-162/global_step162/` if you still have it open. what are the files in that checkpoint?

Here is the list: https://github.com/piegu/language-models/blob/master/images/colab.png

**Comment by stas00** [comment]
Also having problems running your notebook, for some reason I can't load `transformers`, had to force reinstall and run-time restart.
now it's training.

---------------------

OK, based on your snapshot, the file is there, why do you think:
```
file = '/content/test-mlm/checkpoint-162/global_step162/zero_pp_rank_0_mp_rank_00_model_states.pt'
if not os.path.exists(file):
        raise FileNotFoundError(f"can't find '{file}' in directory '{checkpoint_dir}'")
```
fails?

**Comment by stas00** [comment]
ok, the notebook made it to the same failure, I am checking why.

**Comment by piegu** [comment]
> Also having problems running your notebook, for some reason I can't load `transformers`, had to force reinstall and run-time restart.

Yes, that's what I do, too. I do not know why.

**Comment by stas00** [comment]
ok, z2 has a different filenaming convention for saved model states - will fix it shortly

**Comment by stas00** [comment]
can you hold off on changing this notebook - otherwise I lose my changes - I guess I should have forked it first.

**Comment by piegu** [comment]
> I made a fork - otherwise I lose my changes as you apply yours - I guess I should have forked it first.
> 
> To use my branch a simpler way is to:

Thank you. I'm testing.

**Comment by piegu** [comment]
Result
`````Processing zero checkpoint '/content/test-mlm/checkpoint-162/global_step162'
Detected checkpoint of type zero stage 2, world_size: 1

---------------------------------------------------------------------------

TypeError                                 Traceback (most recent call last)

<ipython-input-32-b8e677ab1e2f> in <module>()
      1 from deepspeed.utils.zero_to_fp32 import get_fp32_state_dict_from_zero_chkpt
----> 2 state_dict = get_fp32_state_dict_from_zero_chkpt(checkpoint_dir) # already on cpu
      3 model = model.cpu()
      4 model.load_state_dict(state_dict)
      5 #submit to model hub or save the model to share with others

1 frames

/usr/local/lib/python3.7/dist-packages/deepspeed/utils/zero_to_fp32.py in _get_fp32_state_dict_from_zero_chkpt(ds_checkpoint_dir)
    122         f"Detected checkpoint of type zero stage {zero_stage}, world_size: {world_size}")
    123 
--> 124     model_file = get_model_state_file(ds_checkpoint_dir, zero_stage)
    125     buffers = parse_model_state(model_file)
    126 

TypeError: get_model_state_file() takes 1 positional argument but 2 were given
````

**Comment by stas00** [comment]
Since I can reproduce the problem, please let me complete the process. I will ping you when it's working. Thank you!

**Comment by stas00** [comment]
colab is really a pain to work with - takes forever to install pytorch.

OK, so this doesn't fail any longer:
```
from deepspeed.utils.zero_to_fp32 import get_fp32_state_dict_from_zero_chkpt
state_dict = get_fp32_state_dict_from_zero_chkpt(checkpoint_dir) # already on cpu
model = model.cpu()
model.load_state_dict(state_dict, strict=False)
```
I had to add `, strict=False` as some keys weren't saved:
```
_IncompatibleKeys(missing_keys=['lm_head.decoder.weight', 'lm_head.decoder.bias'], unexpected_keys=[])
```

But it doesn't look right, the expected missing keys don't seem to include `lm_head.decoder.weight`, according to:

https://github.com/huggingface/transformers/blob/dc42e770b86b737251ce5b83f6b0606fe1cd3548/src/transformers/models/roberta/modeling_roberta.py#L1012-L1013

can you please verify that this indeed incorrect?

I meanwhile will replicate your setup on my desktop - don't want to use colab again.

**Comment by stas00** [comment]
I'm pretty sure something is wrong there. I think some state_dict keys got incorrectly saved/loaded. I will try to investigate this and follow up.

**Comment by stas00** [comment]
This proved to be an issue with Alberta model code - https://github.com/huggingface/transformers/issues/12426 - that missing key should be on the list of expected missing keys.

Sorry, I haven't had a chance to actually try the restored model.

So if you do please let me know if it's all dandy and we can then merge this PR.

Thank you.

**Comment by stas00** [comment]
@piegu, please give it a try - I re-coded a big chunk of the foundation as it was quite unstable, and added a lot more checks. Should be producing correct results now. I'm also writing extensive tests on the transformers side.

The issue with the misleading missing key has been addressed as well: https://github.com/huggingface/transformers/pull/12446

So when you test please use `transformers` master.

**Comment by stas00** [comment]
I went ahead and added a very simple to use API (and renamed the previously added one - so my examples above are outdated).

So with this branch now all you need is:

```
from deepspeed.utils.zero_to_fp32 import load_state_dict_from_zero_checkpoint
model = load_state_dict_from_zero_checkpoint(trainer.model, checkpoint_dir)
# now save it and push to hub
# note the model is no longer deepspeed-usable and requires a new deepspeed initialize if you want to continue using it right away
```

File: deepspeed/runtime/engine.py [hunk 1]
```diff
@@ -1960,12 +1963,35 @@ def get_layer_named_buffers(module, prefix=""):
 
         return buffer_names
 
-    def _get_param_shapes(self):
+    def _get_zero_param_shapes(self):
```
Comment (stas00) [inline deepspeed/runtime/engine.py:1966]: @tjruwase, this is the important part that I'd love for you to review.

File: deepspeed/runtime/zero/stage3.py [hunk 2]
```diff
@@ -1191,6 +1202,12 @@ def _create_fp16_partitions_with_defragmentation(self):
 
                 see_memory_usage(f"After Flattening param subgroup {i}", force=False)
 
+        # Perhaps it should assert if any params are unpartitioned.
+        if len(unpartitioned_params) > 0 and torch.distributed.get_rank() == 0:
+            logger.warning(
+                f"{len(unpartitioned_params)} unpartitioned params detected:\n" +
+                "\n".join(unpartitioned_params))
+
```
Comment (stas00) [inline deepspeed/runtime/zero/stage3.py:1210]: I added this as I saw situations where params were missed and weren't partitioned!

I wasn't sure if it should assert or just warn.

the setup is 100 lines earlier.

File: docs/_tutorials/zero.md [hunk 3]
```diff
@@ -270,8 +270,8 @@ You can use this method to save ZeRO-2 weights as well.
 If you'd like to get the fp32 weights, we supply a special script that can do offline consolidation. It requires no configuration files or GPUs. Here is an example of its usage:
 
 ``` bash
-$ cd /path/to/checkpoints_dir
-$ ./zero_to_fp32.py global_step1 pytorch_model.bin
+$ cd /path/to/checkpoint_dir
+$ ./zero_to_fp32.py . pytorch_model.bin
```
Comment (tjruwase) [inline docs/_tutorials/zero.md:274]: Is global_step1 still correct in the next line?

**Review by tjruwase — APPROVED** [review]

File: docs/_tutorials/zero.md (hunk 3 above)
Comment (stas00) [reply docs/_tutorials/zero.md:274]: I changed it to match the `load_checkpoint` algorithm, where only the top path needs to be provided and it then automatically picks the latest checkpoint unless tag is provided (and the script itself doesn't have an option to allow for the tag, but it's trivial to add as the main API supports it).

before that the user had to explicitly pass the checkpoint name.

File: deepspeed/runtime/zero/stage3.py (hunk 2 above)
Comment (stas00) [reply deepspeed/runtime/zero/stage3.py:1210]: For some reason I can't find the original report, but what you're saying answers that concern.

So let me rework/undo this part. Do we want to optionally log that some params are unpartitioned? Perhaps just commenting out this whole added section?

Or I can revert it complete and revisit it down the road if and when there will be a need for it.

I think I will just revert it for now.

Comment (stas00) [reply deepspeed/runtime/zero/stage3.py]: ok, reverted.

Comment (tjruwase) [reply deepspeed/runtime/zero/stage3.py]: Thanks!

**Pull request merged**